src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -3,59 +3,109 @@
 package freechips.rocketchip.diplomacy
 
 import Chisel.{defaultCompileOptions => _, _}
-import freechips.rocketchip.util.CompileOptions.NotStrictInferReset
-import chisel3.{RawModule, MultiIOModule, withClockAndReset, Reset}
-import chisel3.internal.sourceinfo.{SourceInfo, SourceLine, UnlocatableSourceInfo}
+import chisel3.internal.sourceinfo.{SourceInfo, UnlocatableSourceInfo}
+import chisel3.{MultiIOModule, RawModule, Reset, withClockAndReset}
 import freechips.rocketchip.config.Parameters
-import scala.collection.immutable.{SortedMap,ListMap}
+import freechips.rocketchip.util.CompileOptions.NotStrictInferReset
+
+import scala.collection.immutable.{ListMap, SortedMap}
 import scala.util.matching._
 
-abstract class LazyModule()(implicit val p: Parameters)
-{
-  protected[diplomacy] var children = List[LazyModule]()
-  protected[diplomacy] var nodes = List[BaseNode]()
+/** While the [[freechips.rocketchip.diplomacy]] package allows fairly abstract parameter negotiation while constructing a DAG,
+  * [[LazyModule]] builds on top of the DAG annotated with the negotiated parameters and leverage's Scala's lazy evaluation property to split Chisel module generation into two phases:
+  *
+  *   - Phase 1 (diplomatic) states parameters, hierarchy, and connections:
+  *     - [[LazyModule]] and [[BaseNode]] instantiation.
+  *     - [[BaseNode]] binding.
+  *   - Phase 2 (lazy) generates [[chisel3]] Modules:
+  *     - Parameters are negotiated across [[BaseNode]]s.
+  *     - Concrete [[Bundle]]s are created along [[BaseNode]]s and connected
+  *     - [[AutoBundle]] are automatically connected along [[Edges]], punching IO as necessary though module hierarchy
+  *     - [[LazyModuleImpLike]] generates [[chisel3.Module]]s.
+  */
+abstract class LazyModule()(implicit val p: Parameters) {
+  /** Contains sub-[[LazyModule]]s; can be accessed by [[getChildren]]. */
+  protected[diplomacy] var children: List[LazyModule] = List[LazyModule]()
```

Comment:
did you manually add all these types or was that done by an automated process?

💯 for adding them just want to know how paranoid i need to be about them being wrong

Reply:
That was added by Intellij, I double checked it. In order not to be infected by https://stackoverflow.com/questions/52524815/enforce-strict-collections, I make all type strict.
I think I can refactor it back after I get up :p